`all_node_cuts()` performance degrades much faster than predicted by Kanevsky's minimum cut algorithm

NetworkX vs igraph. igraph [implements Kanevsky's too](https://python.igraph.org/en/stable/api/igraph.GraphBase.html#minimum_size_separators).

```python
import timeit

import igraph as ig
import networkx as nx


def bigO(g):
    return (2 ** nx.node_connectivity(g)) * (len(g) ** 3)


def runit(i):
    g = nx.grid_2d_graph(i, i)
    O = bigO(g)
    h = ig.Graph.from_networkx(g)
    t = timeit.timeit(lambda: list(nx.all_node_cuts(g)), number=1)
    t2 = timeit.timeit(lambda: ig.GraphBase.minimum_size_separators(h), number=1)
    return dict(i=i, k=nx.node_connectivity(g), n=len(g), O=O, nx_t=t, ig_t=t2)


if __name__ == "__main__":
    results = []
    for i in range(2, 12):
        r = runit(i)
        results.append(r)
        print(r)
```

Results dataframe w/ percent change calculated for `nx_t` and `ig_t` (* 100, b/c humans). Rate of growth is the interesting bit, b/c NetworkX is 100% Python, and igraph is C w/ a python binding.  Raw data attached [results.json](https://github.com/user-attachments/files/19870471/results.json).

```
    i  k    n        O        nx_t      ig_t  nx_t_pct_chg  ig_t_pct_chg
0   2  2    4      256    0.004447  0.000238           NaN           NaN
1   3  2    9     2916    0.033335  0.000569    649.551704    138.733327
2   4  2   16    16384    0.101915  0.001271    205.725524    123.537768
3   5  2   25    62500    0.143263  0.002748     40.571573    116.194569
4   6  2   36   186624    0.198543  0.006185     38.586098    125.084930
5   7  2   49   470596    0.431762  0.011818    117.465543     91.083902
6   8  2   64  1048576    1.275459  0.022733    195.407745     92.349527
7   9  2   81  2125764    6.076753  0.039182    376.436524     72.357797
8  10  2  100  4000000   32.342393  0.066873    432.231473     70.674632
9  11  2  121  7086244  180.278205  0.109162    457.405282     63.236688
```

Clearly these implementations are following different growth curves. Kanevsky's algorithm should be growing at $\mathcal{O}(2^{k} n^3)$. NetworkX's growth becomes geometric (or worse). 

The cause seems to be for some graph sizes, the number of members of the `antichains(L)` set explodes---billions of members, in some case---but I have not yet figured out why that happens.

_Originally posted by @Cerebus in https://github.com/networkx/networkx/discussions/7989#discussioncomment-12923983_

I put up a property-based test:
````python
@settings(print_blob=True)
@given(
    graph_st(
        _random_connected_graph,
        n=st.integers(5, 10),
        prob=st.floats(0, 1)
    )
)
def test_random_connected_graph(G):
    for flow_func in flow_funcs:
        kwargs = {"flow_func": flow_func}
        edge_cut = nx.minimum_edge_cut(G, **kwargs)
        H = G.copy()
        H.remove_edges_from(edge_cut)
        assert not nx.is_connected(H)
        node_cut = nx.minimum_node_cut(G, **kwargs)
        H = G.copy()
        H.remove_nodes_from(node_cut)
        assert not nx.is_connected(H), f"Failed for {flow_func.__name__} with {G}"
```

Hypothesis is able to find counter-examples. We seem to be strugling with complete graphs:

```
>           assert not nx.is_connected(H), f"Failed for {flow_func.__name__} with {G}"
E           AssertionError: Failed for boykov_kolmogorov with Graph with 5 nodes and 10 edges
E           assert not True
E            +  where True = <function is_connected at 0x7f969999f9c0>(<networkx.classes.graph.Graph object at 0x7f9699dd5850>)
E            +    where <function is_connected at 0x7f969999f9c0> = nx.is_connected
E           Falsifying example: test_random_connected_graph(
E               G=Graph()
E               add_node(0)
E               add_node(1)
E               add_node(2)
E               add_node(3)
E               add_node(4)
E               add_edge(0, 1)
E               add_edge(0, 2)
E               add_edge(0, 3)
E               add_edge(0, 4)
E               add_edge(1, 2)
E               add_edge(1, 3)
E               add_edge(1, 4)
E               add_edge(2, 3)
E               add_edge(2, 4)
E               add_edge(3, 4)
E               ,
E           )
E           Explanation:
E               These lines were always and only run by failing examples:
E                   /home/amcandio/workspace/networkx/networkx/algorithms/connectivity/cuts.py:446
E                   /home/amcandio/workspace/networkx/networkx/algorithms/connectivity/cuts.py:447
E                   /home/amcandio/workspace/networkx/networkx/algorithms/connectivity/cuts.py:610
E                   /home/amcandio/workspace/networkx/networkx/classes/graph.py:436
E                   /home/amcandio/workspace/networkx/networkx/classes/reportviews.py:532
E           
E           You can reproduce this example by temporarily adding @reproduce_failure('6.131.9', b'AEEFKD/wAAAAAAAAQQBBAEEA') as a decorator on your test case
```

For this graph we detect `{1, 2, 3, 4}` as cut, which is wrong

--- a/networkx/algorithms/connectivity/kcutsets.py
+++ b/networkx/algorithms/connectivity/kcutsets.py
@@ -2,7 +2,6 @@
 Kanevsky all minimum node k cutsets algorithm.
 """
 
-import copy
 from collections import defaultdict
 from operator import itemgetter
 
@@ -93,7 +92,6 @@ def all_node_cuts(G, k=None, flow_func=None):
 
     # Address some corner cases first.
     # For complete Graphs
-
     if nx.density(G) == 1:
         yield from ()
         return
@@ -106,9 +104,15 @@ def all_node_cuts(G, k=None, flow_func=None):
     H = build_auxiliary_node_connectivity(G)
     H_nodes = H.nodes  # for speed
     mapping = H.graph["mapping"]
-    # Keep a copy of original predecessors, H will be modified later.
-    # Shallow copy is enough.
-    original_H_pred = copy.copy(H._pred)
+    # The Even-Tarjan reduction requires infinite capacity on external
+    # edges (edges between different original nodes) and capacity 1 on
+    # internal edges (vA -> vB). build_auxiliary_node_connectivity sets
+    # capacity=1 on all edges, which is sufficient for computing the
+    # node connectivity value but not for Kanevsky's algorithm, which
+    # depends on the residual graph structure.
+    for u, w, d in H.edges(data=True):
+        if H_nodes[u]["id"] != H_nodes[w]["id"]:
+            d["capacity"] = float("inf")
     R = build_residual_network(H, "capacity")
     kwargs = {"capacity": "capacity", "residual": R}
     # Define default flow function
@@ -135,25 +139,29 @@ def all_node_cuts(G, k=None, flow_func=None):
         for v in non_adjacent:
             # step 4: compute maximum flow in an Even-Tarjan reduction H of G
             # and step 5: build the associated residual network R
-            R = flow_func(H, f"{mapping[x]}B", f"{mapping[v]}A", **kwargs)
+            # After adding edges in previous iterations, there may be
+            # infinite-capacity paths between x and v, which means no
+            # finite min-cut exists for this pair. Skip in that case.
+            try:
+                R = flow_func(H, f"{mapping[x]}B", f"{mapping[v]}A", **kwargs)
+            except nx.NetworkXUnbounded:
+                continue
             flow_value = R.graph["flow_value"]
 
             if flow_value == k:
-                # Find the nodes incident to the flow.
-                E1 = flowed_edges = [
-                    (u, w) for (u, w, d) in R.edges(data=True) if d["flow"] != 0
-                ]
-                VE1 = incident_nodes = {n for edge in E1 for n in edge}
-                # Remove saturated edges form the residual network.
-                # Note that reversed edges are introduced with capacity 0
-                # in the residual graph and they need to be removed too.
+                # Remove edges with zero residual capacity from R.
+                # Residual capacity = capacity - flow.
+                # For saturated forward edges: cap == flow > 0 -> removed.
+                # For inactive reverse edges: cap == flow == 0 -> removed.
+                # For active reverse edges: cap=0, flow < 0 -> cap != flow
+                #   -> kept (these have positive residual capacity and
+                #   represent cancellable flow in the residual graph).
                 saturated_edges = [
                     (u, w, d)
                     for (u, w, d) in R.edges(data=True)
-                    if d["capacity"] == d["flow"] or d["capacity"] == 0
+                    if d["capacity"] == d["flow"]
                 ]
                 R.remove_edges_from(saturated_edges)
-                R_closure = nx.transitive_closure(R)
                 # step 6: shrink the strongly connected components of
                 # residual flow network R and call it L.
                 L = nx.condensation(R)
@@ -161,62 +169,53 @@ def all_node_cuts(G, k=None, flow_func=None):
                 inv_cmap = defaultdict(list)
                 for n, scc in cmap.items():
                     inv_cmap[scc].append(n)
-                # Find the incident nodes in the condensed graph.
-                VE1 = {cmap[n] for n in VE1}
+                # Compute the transitive closure of L for successor lookups.
+                # Per Picard-Queyranne, each antichain of L corresponds to
+                # a successor-closed set (the antichain plus all successors),
+                # which defines the source side of a minimum s-t cut.
+                L_closure = nx.transitive_closure(L)
                 # step 7: Compute all antichains of L;
                 # they map to closed sets in H.
-                # Any edge in H that links a closed set is part of a cutset.
                 for antichain in nx.antichains(L):
-                    # Only antichains that are subsets of incident nodes counts.
-                    # Lemma 8 in reference.
-                    if not set(antichain).issubset(VE1):
+                    if not antichain:
                         continue
-                    # Nodes in an antichain of the condensation graph of
-                    # the residual network map to a closed set of nodes that
-                    # define a node partition of the auxiliary digraph H
-                    # through taking all of antichain's predecessors in the
-                    # transitive closure.
+                    # Build the successor-closed set in L from the antichain.
+                    S_L = set(antichain)
+                    for scc_node in antichain:
+                        S_L.update(L_closure[scc_node])
+                    # Expand SCC nodes back to R/H nodes.
                     S = set()
-                    for scc in antichain:
-                        S.update(inv_cmap[scc])
-                    S_ancestors = set()
-                    for n in S:
-                        S_ancestors.update(R_closure._pred[n])
-                    S.update(S_ancestors)
+                    for scc_node in S_L:
+                        S.update(inv_cmap[scc_node])
+                    # S must contain the source and not the sink.
                     if f"{mapping[x]}B" not in S or f"{mapping[v]}A" in S:
                         continue
-                    # Find the cutset that links the node partition (S,~S) in H
+                    # Find the cutset: edges from S to ~S in H.
                     cutset = set()
                     for u in S:
-                        cutset.update((u, w) for w in original_H_pred[u] if w not in S)
-                    # The edges in H that form the cutset are internal edges
-                    # (ie edges that represent a node of the original graph G)
-                    if any(H_nodes[u]["id"] != H_nodes[w]["id"] for u, w in cutset):
+                        cutset.update((u, w) for w in H[u] if w not in S)
+                    if not cutset:
                         continue
                     node_cut = {H_nodes[u]["id"] for u, _ in cutset}
 
-                    if len(node_cut) == k:
-                        # The cut is invalid if it includes internal edges of
-                        # end nodes. The other half of Lemma 8 in ref.
-                        if x in node_cut or v in node_cut:
-                            continue
-                        if node_cut not in seen:
-                            yield node_cut
-                            seen.append(node_cut)
+                    if node_cut not in seen:
+                        yield node_cut
+                        seen.append(node_cut)
 
                 # Add an edge (x, v) to make sure that we do not
                 # find this cutset again. This is equivalent
                 # of adding the edge in the input graph
                 # G.add_edge(x, v) and then regenerate H and R:
                 # Add edges to the auxiliary digraph.
+                # External edges get infinite capacity.
+                H.add_edge(f"{mapping[x]}B", f"{mapping[v]}A", capacity=float("inf"))
+                H.add_edge(f"{mapping[v]}B", f"{mapping[x]}A", capacity=float("inf"))
+                # Add edges to the residual network.
                 # See build_residual_network for convention we used
                 # in residual graphs.
-                H.add_edge(f"{mapping[x]}B", f"{mapping[v]}A", capacity=1)
-                H.add_edge(f"{mapping[v]}B", f"{mapping[x]}A", capacity=1)
-                # Add edges to the residual network.
-                R.add_edge(f"{mapping[x]}B", f"{mapping[v]}A", capacity=1)
+                R.add_edge(f"{mapping[x]}B", f"{mapping[v]}A", capacity=R.graph["inf"])
                 R.add_edge(f"{mapping[v]}A", f"{mapping[x]}B", capacity=0)
-                R.add_edge(f"{mapping[v]}B", f"{mapping[x]}A", capacity=1)
+                R.add_edge(f"{mapping[v]}B", f"{mapping[x]}A", capacity=R.graph["inf"])
                 R.add_edge(f"{mapping[x]}A", f"{mapping[v]}B", capacity=0)
 
                 # Add again the saturated edges to reuse the residual network

--- a/networkx/algorithms/connectivity/tests/test_kcutsets.py
+++ b/networkx/algorithms/connectivity/tests/test_kcutsets.py
@@ -8,7 +8,7 @@ import networkx as nx
 from networkx.algorithms import flow
 from networkx.algorithms.connectivity.kcutsets import _is_separating_set
 
-MAX_CUTSETS_TO_TEST = 4  # originally 100. cut to decrease testing time
+MAX_CUTSETS_TO_TEST = 100
 
 flow_funcs = [
     flow.boykov_kolmogorov,
@@ -133,7 +133,6 @@ def _check_separating_sets(G):
             assert not nx.is_connected(nx.restricted_view(G, cut, []))
 
 
-@pytest.mark.slow
 def test_torrents_and_ferraro_graph():
     G = torrents_and_ferraro_graph()
     _check_separating_sets(G)
@@ -207,7 +206,6 @@ def test_disconnected_graph():
     pytest.raises(nx.NetworkXError, next, cuts)
 
 
-@pytest.mark.slow
 @pytest.mark.parametrize("G", [nx.grid_2d_graph(4, 4), nx.cycle_graph(5)])
 @pytest.mark.parametrize("flow_func", flow_funcs)
 def test_alternative_flow_functions(G, flow_func):

